### ISSUE TITLE
proxy: allow hot-reloading of policy on change

### ISSUE THREAD

**Issue by yegle** [pomerium/pomerium#89]
Right now when you edit `policy.yaml`, you have to restart Pomerium (at least for users who run Pomerium in docker).

It would be great if Pomerium can monitor this file and trigger a reload upon file change (or with a given interval).

Ideally Pomerium should only reload when the semantic of the policy file is changed. E.g. when someone add a comment line, it should not cause a reload. But if it's too hard I'm fine with any change (e.g. compare the hash with an interval).

**Comment by victornoel** [comment]
For the record, this also makes sense in Kubernetes where `ConfigMap` can be updated independently of `Deployments`.

That being said, when deploying pomerium with Helm, it is possible to use the [`checksum` annotation tricks](https://helm.sh/docs/developing_charts/#automatically-roll-deployments-when-configmaps-or-secrets-change) to force the redeployment of a pod when another resources (such as a `ConfigMap`) changes.

**Comment by travisgroth** [comment]
@desimone I've done some looking at this as I have interest seeing this go a bit further.  Long term vision would be a kube ingress controller or operator.  Transparent updates of policy (at least) become important for that kind of model.  If we scope the "hot" changes to just policy, three things need to happen:

* Add new methods to Authorize and Proxy to re-read and atomically swap out the internal data structures derived from the policy file.
* Add a utility method to watch the policy file (the ENV var obviously is static).
* Add a watcher goroutine in the main package (or each service?) to call the policy update methods on relevant service structs.  

If what I've outlined sounds rational, I think I can put together a PR.

I have a slightly related question after looking around the code for this change - how committed to using pomerium/envconfig and using parsing logic in the service structs are you?  I'm inclined to switch to viper and move all the config - including policy - into a single config file.  This would remove a lot of loading/validation code and should retain ENV sourcing for everything if desired.

This would be a (slight) breaking change in the config file format and probably make sense to rename the file.  If there's no interest or you have a specific objection, I'll leave it be, but I can open a different issue on this if there's discussion to be had.  It would alleviate having to write the watcher method but otherwise doesn't significantly alter this feature.

**Comment by desimone** [comment]
@travisgroth Thank you for taking a look at this! Hot reloading of changes is absolutely I'd like to see supported but haven't had a chance to think carefully about it.


> -Add new methods to Authorize and Proxy to re-read and atomically swap out the internal data structures derived from the policy file.
> -Add a utility method to watch the policy file (the ENV var obviously is static).
> -Add a watcher goroutine in the main package (or each service?) to call the policy update methods on relevant service structs.

That sounds right right. The only additional edge case I can think of right now is making sure that when an update occurs, we guard against unexpected situations where only N of X nodes may have received and processed the update. 

> how committed to using pomerium/envconfig and using parsing logic in the service structs are you

I'm open to using using viper so long as the env_vars are still supported (for those who want to use them) and breaking changes are minimal. Could you give me a sense of what the breaking changes would be?

**Comment by travisgroth** [comment]
>That sounds right right. The only additional edge case I can think of right now is making sure that when an update occurs, we guard against unexpected situations where only N of X nodes may have received and processed the update.

The approach I was looking at protects against a partial policy load in a given instance.  To introduce coordination on policy "version" would be quite a bit more complicated.  What would the desired behavior be?  

Across different service layers we could maybe attach a config hash and require the policy be identical between authz and proxy when we process a request...though that means transient user impact whenever you make a change.  I'm not sure that's worth it.

Between instances of a given service, we'd almost have to use a distributed k/v to do any coordination.  Maybe just make the scenario easy to debug/identify and make sure it's clear what policy is loaded in the logs?  If we have that checksum calculated we can also expose it via prometheus (another issue I may try to take a swing at later), so at least your monitoring system could tell you about an uncoordinated policy?  As I'm typing, that seems like an easy first iteration.

>Could you give me a sense of what the breaking changes would be?

Basically, policy becomes a key under the top level yaml.  Sample config would look like this:

```yaml
authenticate_service_url: https://authenticate.corp.com
authorize_service_url: https://authorize.corp.com
policy:
  - from: neverssl.corp.com
     to: http://neverssl.com
     allowed_domains:
      - pomerium.io
```
I tried using viper to parse the policy as-is and it doesn't seem like it can handle a top level list, even if we wanted to take that path.  I can create an issue for this to discuss further.

**Comment by desimone** [comment]
> Maybe just make the scenario easy to debug/identify and make sure it's clear what policy is loaded in the logs?

 For now yes. 

> If we have that checksum calculated we can also expose it via prometheus (another issue I may try to take a swing at later), so at least your monitoring system could tell you about an uncoordinated policy? As I'm typing, that seems like an easy first iteration.

That’s exactly what I was thinking as well.  We can tackle the harder bits in a subsequent release. 

> Basically, policy becomes a key under the top level yaml.

I’m for it given what we get in return. 👍🏼

**Comment by yegle** [comment]
Just wondering if Pomerium will hot-reload certificate as well? My current certificate has not expired yet, and I'm not sure if I need to restart the process once the certificate is renewed.

**Comment by travisgroth** [comment]
@yegle It does not.  

My thought was policy changes would be much more frequent so that would have the highest payoff.  With certs you run into some complexity with hot reloading since you have multiple files to synchronize.  I would actually lean towards saying that cert reload should only be done on SIGHUP (or maybe a reload endpoint) so the user can make the change atomic and predictable.

Reinitializing more than the policy data structures probably means a strategy to just throw the entire set of services away and reinitialize as if the process was restarted.  Doing so safely (so that bad changes don't cause a crash) will require extra diligence in the validation code.  Can you open a new issue to track this more complex feature?
